File: source/client/src/mappings/AutoMapperProfile.ts
//
// Copyright (c) 2023-2024 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/zenn-search/blob/main/LICENSE
//

import {
  createMap,
  createMapper,
  forMember,
  mapFrom
} from '@automapper/core';
import { pojos, PojosMetadataMap } from '@automapper/pojos';

import {
  IndexDataDTO,
  SearchIndexData,
  SearchIndexDataDTO
} from '../types/Model';

export const mapper = createMapper({
  strategyInitializer: pojos()
});

PojosMetadataMap.create<IndexDataDTO>('IndexDataDTO', {
  title: String,
  content: String,
  created: String,
  updated: String,
  etag: String
});

PojosMetadataMap.create<SearchIndexDataDTO>('SearchIndexDataDTO', {
  '@search.score': Number,
  id: String,
  value: 'IndexDataDTO'
});

PojosMetadataMap.create<SearchIndexData>('SearchIndexData', {
  score: Number,
  id: String,
  title: String,
  content: String,
  published: String
});

createMap<SearchIndexDataDTO, SearchIndexData>(
  mapper,
  'SearchIndexDataDTO',
  'SearchIndexData',
  forMember((target) => target.score, mapFrom((source) => source['@search.score'])),
  forMember((target) => target.title, mapFrom((source) => source.value?.title)),
  forMember((target) => target.emoji, mapFrom((source) => source.value?.emoji)),
  forMember((target) => target.content, mapFrom((source) => source.value?.content)),
  forMember(
    (target) => target.published,
    mapFrom((source) => source.id.replace(/(\d{4})_(\d{2})_(\d{2})_(\d{2})(\d{2})(\d{2})/, '$1/$2/$3 $4:$5')))
);

File: source/server/Karamem0.ZennSearch.Functions/Mappings/AutoMapperProfile.cs
//
// Copyright (c) 2023-2024 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/zenn-search/blob/main/LICENSE
//

using AutoMapper;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;

namespace Karamem0.ZennSearch.Models.Mappings;

public class AutoMapperProfile : Profile
{

    public static Assembly Assembly => typeof(AutoMapperProfile).Assembly;

    public AutoMapperProfile()
    {
        _ = this.CreateMap<IndexData, SearchIndexData>()
            .ForMember(d => d.Value, o => o.MapFrom(s => s));
        _ = this.CreateMap<IndexData, SearchIndexItemData>();
    }

}
